### PR TITLE
Updated comments for all the functions and added a doc comment to the pa...

### PR DESCRIPTION
**Pull request by btracey**
...ckage

### CODE REVIEW

File: native/doc.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
+/*
+Package native is a Go implementation of the BLAS API. This implementation
+panics when the input arguments are invalid as per the standard, for example
+if a vector increment is zero. Please note that the treatment of NaN values
+is not specified, and differs among the BLAS implementations.
+github.com/gonum/blas/blas64 provides helpful wrapper functions to the BLAS
+interface. The rest of this text describes the layout of the data for the input types.
+Please note that in the function documentation, x[i] refers to the i^th element
```
Comment (kortschak) [inline native/doc.go]: New line.

File: native/doc.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+of the vector, which will be different from the i^th element of the slice if
+incX != 1.
+
+See http://www.netlib.org/lapack/explore-html/d4/de1/_l_i_c_e_n_s_e_source.html
+for more license information.
+
+Vector arguments are effectively strided slices. They have two input arguments,
+a number of elements, n, and an increment, incX. The increment specifies the
+distance between elements of the vector. The actual go slice may be longer
```
Comment (kortschak) [inline native/doc.go]: s/go/Go/

File: native/doc.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+[a00 ... a0n a0* ... a1stride-1 a21 ... amn am* ... amstride-1]
+Thus, dense[i*ld + j] refers to the {i, j}th element of the matrix.
+
+Symmetric and triangular matrices (non-packed) are stored identically to Dense,
+except only elements in one triangle of the matrix are accessed.
+
+Packed symmetric and packed triangular matrices are laid out with the entries
+condensed such that all of the zeros are removed. So, the upper triangular
+matrix
+[
```
Comment (kortschak) [inline native/doc.go]: Indent the brackets too - also below.

File: native/doc.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+necessary. So, for example, if n = 5 and incX = 3, the elements of s are
+	[0 * * 1 * * 2 * * 3 * * 4 * * * ...]
+where * elements are never accessed. If incX = -3, the same elements are
+accessed, just in reverse order (4, 3, 2, 1, 0).
+
+Dense matrices are specified by a number of rows, a number of columns, and a stride.
+The stride specifies the number of entries in the slice between the first element
+of successive rows. The stride must be at least as large as the number of columns
+but may be longer.
+[a00 ... a0n a0* ... a1stride-1 a21 ... amn am* ... amstride-1]
```
Comment (kortschak) [inline native/doc.go]: Indent.

File: native/doc.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+Thus, dense[i*ld + j] refers to the {i, j}th element of the matrix.
+
+Symmetric and triangular matrices (non-packed) are stored identically to Dense,
+except only elements in one triangle of the matrix are accessed.
```
Comment (kortschak) [inline native/doc.go]: s/except only/except that only/

File: native/doc.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+Packed symmetric and packed triangular matrices are laid out with the entries
+condensed such that all of the zeros are removed. So, the upper triangular
```
Comment (kortschak) [inline native/doc.go]: This is misleading. It suggests some magical packing where zero elements that are in the used triangle/half are not stored but reintroduced when needed. I can't think of better wording though.

File: native/doc.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+[
+   *  1  2  3
+   4  5  6  7
+   8  9 10 11
+  12 13 14 15
+  16 17 18  *
+  19 20  *  *
+]
+which is given to the BLAS routine is [* 1 2 3 4 ...].
+See http://www.crest.iu.edu/research/mtl/reference/html/banded.html
```
Comment (kortschak) [inline native/doc.go]: New line.

File: native/level2double.go [hunk 8]
```diff
@@ -247,10 +195,12 @@ func (Implementation) Dger(m, n int, alpha float64, x []float64, incX int, y []f
...
 }
 
-// Dgbmv performs y = alpha*A*x + beta*y where a is an mxn band matrix with
-// kl subdiagonals and ku super-diagonals. m and n refer to the size of the full
-// dense matrix, not the actual number of rows and columns in the contracted banded
-// matrix.
+// Dgbmv computes
+//  y = alpha * A * x + beta * y if tA == blas.NoTrans
+//  y = alpha * A^T * x + beta * y if tA == blas.Trans
+// where a is an mxn band matrix kl subdiagonals and ku super-diagonals, and
```
Comment (kortschak) [inline native/level2double.go]: s/ku/kU/ s/kl/kL/

File: native/dgemm.go [hunk 9]
```diff
@@ -19,11 +19,10 @@ const (
...
 )
 
-// Dgemm computes c := beta * C + alpha * A * B. If tA or tB is blas.Trans,
-// A or B is transposed.
-// m is the number of rows in A or A transpose
-// n is the number of columns in B or B transpose
-// k is the columns of A and rows of B
+// Dgemm computes
+//  C = beta * C + alpha * A * B.
+// tA and tB specify if A or B are transposed. A, B, and C are all n x n dense
```
Comment (kortschak) [inline native/dgemm.go]: s/if/whether/ ?

File: native/level1double.go [hunk 10]
```diff
@@ -79,9 +60,9 @@ func (Implementation) Ddot(n int, x []float64, incX int, y []float64, incY int)
 	return asm.DdotInc(x, y, uintptr(n), uintptr(incX), uintptr(incY), uintptr(ix), uintptr(iy))
 }
 
-// Dnrm2 computes the euclidean norm of a vector, sqrt(x'x).
-// This function returns 0 if the increment is negative. This behavior matches
-// the reference implementation.
+// Dnrm2 computes the euclidean norm of a vector,
```
Comment (kortschak) [inline native/level1double.go]: s/euclidean/Euclidean/

File: native/level1double.go [hunk 11]
```diff
@@ -312,16 +297,15 @@ func (Implementation) Daxpy(n int, alpha float64, x []float64, incX int, y []flo
 	asm.DaxpyInc(alpha, x, y, uintptr(n), uintptr(incX), uintptr(incY), uintptr(ix), uintptr(iy))
 }
 
-// Drotg gives plane rotation
-//
+// Drotg computes the plane rotation
 // _      _    _   _     _   _
```
Comment (kortschak) [inline native/level1double.go]: Indent.

File: native/level1double.go [hunk 12]
```diff
@@ -312,16 +297,15 @@ func (Implementation) Daxpy(n int, alpha float64, x []float64, incX int, y []flo
...
-// r = ±(a^2 + b^2)
-// c = a/r, the cosine of the plane rotation
-// s = b/r, the sine of the plane rotation
+// where
+//  r = ±(a^2 + b^2)
+//  c = a/r, the cosine of the plane rotation
+//  s = b/r, the sine of the plane rotation
 //
 // NOTE: Netlib reference implementation returns a different
 // sign for r when a or b is zero than the BLAS technical manual. Other
```
Comment (kortschak) [inline native/level1double.go]: "than"?

If this an Americanism?

"When a or b is zero, the Netlib reference implementation returns a different sign for r compared with the BLAS technical manual."

In English we use to, "compared with" avoids the issue.

File: native/level2double.go [hunk 13]
```diff
@@ -854,10 +806,10 @@ func (Implementation) Dsymv(ul blas.Uplo, n int, alpha float64, a []float64, lda
 	}
 }
 
-// Dtbmv  performs one of the matrix-vector operations
-// 		x := A*x,   or   x := A**T*x,
-// where x is an n element vector and  A is an n by n unit, or non-unit,
-// upper or lower triangular band matrix.
+// Dtbmv computes
+// 	x = A * x if tA == blas.NoTrans
+//  x = A^T * x if tA == blas.Trans
```
Comment (kortschak) [inline native/level2double.go]: Choose an indentation level.

File: native/level2double.go [hunk 14]
```diff
@@ -1055,10 +1007,10 @@ func (Implementation) Dtbmv(ul blas.Uplo, tA blas.Transpose, d blas.Diag, n, k i
 	}
 }
 
-// Dtpmv performs one of the matrix-vector operations
-// 		x := A*x,   or   x := A**T*x,
-// where x is an n element vector and  A is an n by n unit, or non-unit,
-// upper or lower triangular matrix represented in packed storage format.
+// Dtpmv computes
+// 	x = A * x if tA == blas.NoTrans
+//  x = A^T * x if tA == blas.Trans
```
Comment (kortschak) [inline native/level2double.go]: Indentation is inconsistent.

**Comment by kortschak** [comment]
General comment - always run a local `godoc -http=:6060` and look at the rendering.

Where you use "mxn", maybe consider "m⨉n" or "m by n". Similarly for "n x n". Harmonise spacing and use of "by", "x" or "⨉".

For Drotm, s/2 x N/2 x n/ with above consideration.

File: native/dgemm.go (hunk 9 above)
Comment (btracey) [inline native/dgemm.go]: Done

File: native/doc.go (hunk 1 above)
Comment (btracey) [inline native/doc.go]: Done

File: native/doc.go (hunk 2 above)
Comment (btracey) [inline native/doc.go]: Done

File: native/doc.go (hunk 4 above)
Comment (btracey) [inline native/doc.go]: Done

File: native/doc.go (hunk 5 above)
Comment (btracey) [inline native/doc.go]: Done.

File: native/doc.go (hunk 6 above)
Comment (btracey) [inline native/doc.go]: So that all of the unused elements are removed?

File: native/doc.go (hunk 3 above)
Comment (btracey) [inline native/doc.go]: Done.

File: native/doc.go (hunk 7 above)
Comment (btracey) [inline native/doc.go]: Done.

File: native/level1double.go (hunk 10 above)
Comment (btracey) [inline native/level1double.go]: Done.

File: native/level1double.go (hunk 11 above)
Comment (btracey) [inline native/level1double.go]: Done.

File: native/level1double.go (hunk 12 above)
Comment (btracey) [inline native/level1double.go]: I guess it's an americanism. "X has a different sign than Y" sounds perfectly natural to me. "X has a different sign to Y" sounds weird.

Used compared with.

File: native/level2double.go (hunk 8 above)
Comment (btracey) [inline native/level2double.go]: Done.

File: native/level2double.go (hunk 13 above)
Comment (btracey) [inline native/level2double.go]: Done. Sorry.

File: native/level2double.go (hunk 14 above)
Comment (btracey) [inline native/level2double.go]: Done.

**Comment by btracey** [comment]
Okay, I think I got all of the n⨉n rectified.

**Comment by btracey** [comment]
I also subbed ∗ for * in non-indented blocks for elements that are not accessed. It looks much better in godoc.

**Comment by btracey** [comment]
PTAL

File: native/level1double.go (hunk 12 above)
Comment (kortschak) [inline native/level1double.go]: The sentence structure should change here too. The current order makes it unnecessarily difficult to parse. This is fixed by either leading with the conditional or finishing with the conditional. I led, but the other order is fine too.

**Comment by btracey** [comment]
Fixed Drotg comment. PTAL.

File: native/doc.go (hunk 6 above)
Comment (kortschak) [inline native/doc.go]: "unreferenced", "redundant"?

**Comment by kortschak** [comment]
s/on/regarding/ - this may be another language variant conflict.

s/. This agrees with/ and/

File: native/doc.go (hunk 6 above)
Comment (btracey) [inline native/doc.go]: I like unreferenced. Fixed.

**Comment by btracey** [comment]
Done.

**Comment by kortschak** [comment]
It looks like '⨉' is not available on all machines (e.g. not on my workstation). Either try '⨯' (named "Vector or Cross Product"), '×' (named "Multiplication sign" - low unicode point, so probably more available), or revert to 'x'. Sorry.

Otherwise LGTM, thank you.

**Pull request merged**